pkg/core/blockchain.go

```diff
@@ -732,31 +753,50 @@ func (bc *Blockchain) resetStateInternal(height uint32, stage stateChangeStage)
 			blocksCnt, batchCnt, keysCnt int
 		)
 		for i := height + 1; i <= currHeight; i++ {
-			err := cache.DeleteBlock(bc.GetHeaderHash(int(i)))
+			err = cache.DeleteBlock(bc.GetHeaderHash(int(i)))
 			if err != nil {
-				return fmt.Errorf("error while removing block %d: %w", i, err)
+				// It's a critical error, stop the cycle after it.
+				err = fmt.Errorf("error while removing block %d: %w", i, err)
+				break
 			}
 			blocksCnt++
 			if blocksCnt == persistBatchSize {
-				keys, err = cache.Persist()
-				if err != nil {
-					return fmt.Errorf("failed to persist intermediate batch removed blocks, transactions and AERs: %w", err)
-				}
 				blocksCnt = 0
 				batchCnt++
-				keysCnt += keys
-				bc.log.Info("intermediate batch of removed blocks, transactions and AERs is persisted", zap.Int("batches persisted", batchCnt), zap.Duration("took", time.Since(p)), zap.Int("keys", keys))
+				batchPersistCh <- persistData{
+					postPersist: func(persistedKeys int, err error) {
+						if err != nil {
+							// Need to wait the exit routine to finish.
+							// At the same time, continue execution and do not treat this error as critical,
+							// changes still have the possibility to be persisted later before OOM killer comes.
```

Comment:
Most likely it'll fail again, so it'd be safer to just stop. Persist errors are things like running out of disk space, you can wait/retry in general, but it's easier to stop, restart can be handled normally anyway.

Reply:
Fixed.